### PR TITLE
added additional loss against data for NNODE

### PR DESCRIPTION
**Pull request by AstitvaAggarwal**
added option for additional loss against data for NNODE(used PhysicsInformedNN as a reference),work in progress as im getting some errors would greatly appreciate any help. Thanks
. (trying to solve issue #640, but NNODE dosent allow additional losses against surrogate Model predictions nor does it allow parameter estimation so this pr could help solve such ODE problems)

### CODE REVIEW

**Comment by xtalax** [comment]
This looks good, can you write some tests to ensure its working? We will need to do documentation too.

**Comment by AstitvaAggarwal** [comment]
in the paper the loss term has additional loss terms :Lgls and Lconstr(under introduction eq 3 to 8), Lgls is defined against cell density predictions via a Surrogate Model MLP which takes in training data as measured data(x,t), Lconstr performs in parameter Estimation for the Growth, Time Delay and Diffusion terms(parameter functions are functions of the solution itself -> leads to issue #572).

**Comment by xtalax** [comment]
Ah I see it, yep this looks like it implements what's missing, good job (y)

**Comment by ChrisRackauckas** [comment]
The test failure looks real.

File: src/ode_solve.jl [hunk 1]
```diff
@@ -283,10 +291,44 @@ function generate_loss(strategy::StochasticTraining, phi, f, autodiff::Bool, tsp
...
+    weights = strategy.weights ./ sum(strategy.weights)
+
+    N = length(weights)
+    samples = strategy.samples
+
+    difference = (maxT - minT) / N
+
+    data = Float64[]
+    for (index, item) in enumerate(weights)
+        temp_data = rand(1, trunc(Int, samples * item)) .* difference .+ minT .+
```
Comment (ChrisRackauckas) [inline src/ode_solve.jl:312]: Rebase issue? Rebase to the new master.

**Comment by ChrisRackauckas** [comment]
Rebase this onto the new master so that it's not dependent on any other PR (it's now merged).

File: src/ode_solve.jl [hunk 2]
```diff
@@ -407,7 +418,34 @@ function DiffEqBase.__solve(prob::DiffEqBase.AbstractODEProblem,
...
+    elseif strategy isa StochasticTraining
+        Optimization.AutoZygote()
+    elseif strategy isa WeightedIntervalTraining
+        Optimization.AutoZygote()
+    else
+        # by default GridTraining choice of Optimization
+        # if adding new training algorithms we can extend this,
+        # if-elseif-else block for choices of optimization algos
+        Optimization.AutoZygote()
+    end
```
Comment (ChrisRackauckas) [inline src/ode_solve.jl:445]:
```suggestion
    opt_algo = if strategy isa QuadratureTraining
        Optimization.AutoForwardDiff()
    else
        Optimization.AutoZygote()
    end
```

File: src/ode_solve.jl [hunk 3]
```diff
@@ -23,7 +23,14 @@ of the physics-informed neural network which is used as a solver for a standard
...
 ## Keyword Arguments
-
+* `additional_loss`: A function additional_loss(phi, θ) where phi are the neural network trial solutions,
+                     θ are the weights of the neural network(s).
+example: 
+    ts=[t for t in 1:100]
+    (u_, t_) = (analytical_func(ts), ts)
+    function additional_loss(phi, θ)
+        return sum(sum(abs2, [phi(t, θ) for t in t_] .- u_)) / length(u_)
+    end
```
Comment (ChrisRackauckas) [inline src/ode_solve.jl:33]: This isn't in the right spot. Make an example section.

File: src/ode_solve.jl [hunk 4]
```diff
@@ -312,7 +322,8 @@ function generate_loss(strategy::WeightedIntervalTraining, phi, f, autodiff::Boo
             sum(abs2, [inner_loss(phi, f, autodiff, t, θ, p) for t in ts])
         end
     end
-    optf = OptimizationFunction(loss, Optimization.AutoZygote())
+
+    return loss
```
Comment (ChrisRackauckas) [inline src/ode_solve.jl:326]:
```suggestion
    return loss
```

File: src/ode_solve.jl [hunk 5]
```diff
@@ -250,38 +260,38 @@ function generate_loss(strategy::QuadratureTraining, phi, f, autodiff::Bool, tsp
...
-
         if batch
             sum(abs2, inner_loss(phi, f, autodiff, ts, θ, p))
         else
             sum(abs2, [inner_loss(phi, f, autodiff, t, θ, p) for t in ts])
         end
     end
-    optf = OptimizationFunction(loss, Optimization.AutoZygote())
+
+    return loss
```
Comment (ChrisRackauckas) [inline src/ode_solve.jl:294]:
```suggestion
    return loss
```

File: src/ode_solve.jl (hunk 3 above)
Comment (AstitvaAggarwal) [reply src/ode_solve.jl:33]: under this argument's description? or along with the lower example?

**Review by ChrisRackauckas — APPROVED** [review]
